hist/histv7/inc/ROOT/RHistImpl.hxx

```diff
@@ -227,122 +257,467 @@ namespace Internal {
-         hist->GrowAxis(thisAxis, x[thisAxis]);
-         status = RAxisBase::EFindStatus::kCanGrow;
+      return total_regular_bins_before;
+   }
+};
 
-         // Abort bin calculation; we don't care. Let RHist::GetBinIndex() retry!
-         return bin;
+template <int I, int NDIMS, typename BINS, class AXES>
+struct RComputeGlobalBin {
+   int operator()(int total_regular_bins_before, const AXES &axes, const BINS &virtualBins, const BINS &binSizes, const BINS &localBins) const
+   {
+      // We can tell how many regular bins lie before us on this axis,
+      // accounting for the underflow bin of this axis if it has one.
+      const int num_underflow_bins = static_cast<int>(!std::get<I>(axes).CanGrow());
+      const int num_regular_bins_before =
+         std::max(virtualBins[I] - num_underflow_bins, 0);
+      total_regular_bins_before += num_regular_bins_before * binSizes[I];
+
+      // If we are on an overflow or underflow bin on this axis, we know that
+      // we don't need to look at the remaining axes. Projecting on those
+      // dimensions would only take us into an hyperplane of over/underflow
+      // bins for the current axis, and we know that by construction there
+      // will be no regular bins in there.
+      if (localBins[I] < 1)
+         return total_regular_bins_before;
+      
+      return RComputeGlobalBin<I - 1, NDIMS, BINS, AXES>()(total_regular_bins_before, axes, virtualBins, binSizes, localBins);
+   }
+};
+
+/// Recursively compute some quantities needed for `ComputeLocalBins`, namely
+/// the total number of bins per hyperplane (overflow and regular) and the
+/// number of regular bins per hyperplane on the hyperplanes that have them.
+template <int I, int NDIMS, class AXES>
+struct RComputeLocalBinsInitialisation;
+
+template <int NDIMS, class AXES>
+struct RComputeLocalBinsInitialisation<0, NDIMS, AXES> {
+   void operator()(std::array<int, NDIMS-1> /* bins_per_hyperplane */, std::array<int, NDIMS-1> /* regular_bins_per_hyperplane */, const AXES & /*axes*/) const
+   {}
+};
+
+template <int I, int NDIMS, class AXES>
+struct RComputeLocalBinsInitialisation {
+   void operator()(std::array<int, NDIMS-1>& bins_per_hyperplane, std::array<int, NDIMS-1>& regular_bins_per_hyperplane, const AXES &axes) const
+   {
+      constexpr const int thisAxis = NDIMS - I - 1;
+      bins_per_hyperplane[thisAxis] = Internal::RGetNBinsCount<thisAxis, AXES>()(axes);
+      regular_bins_per_hyperplane[thisAxis] = Internal::RGetNBinsNoOverCount<thisAxis, AXES>()(axes);
+      RComputeLocalBinsInitialisation<I - 1, NDIMS, AXES>()(bins_per_hyperplane, regular_bins_per_hyperplane, axes);
+   }
+};
+
+/// Recursively computes the number of regular bins before the current dimension,
+/// as well as the number of under- and overflow bins left to account for after
+/// the current dimension.
+/// If the latter is equal to 0, there is no need to process further.
```

Comment:
As for `RComputeGlobalBin` above, I think spelling out some extra context could make this comment clearer:

- We're currently looking at an underflow/overflow bin.
- We're working in global coordinates, so our index is a count of previous overflow bins.
- We want to know how many _regular_ bins lie before us.
- We proceed recursively by starting at the highest dimension, examining how many full hyperplanes of regular bins lie before us, then projecting on the remaining dimensions.